lib/tests/streamlit/runtime/state/query_params_test.py

```diff
@@ -0,0 +1,138 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.state.query_params import QueryParams
+from tests.delta_generator_test_case import DeltaGeneratorTestCase
+
+
+class QueryParamsMethodTests(DeltaGeneratorTestCase):
+    def setUp(self):
+        super().setUp()
+        self._query_params = QueryParams()
+        # avoid using ._query_params as that will use __setattr__
+        self._query_params.__dict__["_query_params"] = {"foo": "bar", "two": ["x", "y"]}
+
+    def test__getitem__raises_KeyError_for_nonexistent_key(self):
+        with pytest.raises(KeyError):
+            self._query_params["nonexistent"]
+
+    def test__getattr__raises_AttributeError_for_nonexistent_key(self):
+        with pytest.raises(AttributeError):
+            self._query_params.nonexistent
+
+    def test__getitem__returns_last_element_of_list(self):
+        assert self._query_params["two"] == "y"
+
+    def test__getitem__converts_int_value_to_string(self):
+        self._query_params["baz"] = 1
+        assert self._query_params["baz"] == "1"
+
+    def test__getitem__converts_float_value_to_string(self):
+        self._query_params["corge"] = 1.23
+        assert self._query_params["corge"] == "1.23"
+
+    def test__getitem__retrieves_existing_key(self):
+        assert self._query_params["foo"] == "bar"
+
+    def test_get_method_retrieves_existing_key(self):
+        assert self._query_params.get("foo") == "bar"
+
+    def test_get_method_returns_default_for_nonexistent_key(self):
+        return_val = self._query_params.get("dog", default="bark")
+        assert return_val == "bark"
```

Comment:
These tests don't seem useful to me because we're really just testing the implementation of `MutableMapping`. It'd be more useful to just drop all tests of auto-implemented methods and add a test checking that `isinstance(self._query_params, MutableMapping)`